tests/handlers/test_users.py

```diff
@@ -74,17 +74,72 @@ def test_returns_400_when_email_taken_case_insensitive(self):
 
 
 class TestUserListGet(BaseTestCase):
+    def make_request_and_check_users(self, method, path, expected_users, unexpected_users, *args, **kwargs):
+        rv = self.make_request(method, path, *args, **kwargs)
+        user_ids = map(lambda u: u['id'], rv.json['results'])
+        for user in expected_users:
+            self.assertIn(user.id, user_ids)
+        for user in unexpected_users:
+            self.assertNotIn(user.id, user_ids)
+
     def test_returns_users_for_given_org_only(self):
         user1 = self.factory.user
         user2 = self.factory.create_user()
         org = self.factory.create_org()
         user3 = self.factory.create_user(org=org)
 
-        rv = self.make_request('get', "/api/users")
-        user_ids = map(lambda u: u['id'], rv.json['results'])
-        self.assertIn(user1.id, user_ids)
-        self.assertIn(user2.id, user_ids)
-        self.assertNotIn(user3.id, user_ids)
+        self.make_request_and_check_users('get', '/api/users', [user1, user2], [user3])
+
+    def test_returns_users_by_filter(self):
+        now = models.db.func.now()
+        user_enabled_active1 = self.factory.create_user(disabled_at=None, is_invitation_pending=None)
+        user_enabled_active2 = self.factory.create_user(disabled_at=None, is_invitation_pending=False)
+        user_enabled_pending = self.factory.create_user(disabled_at=None, is_invitation_pending=True)
+        user_disabled_active1 = self.factory.create_user(disabled_at=now, is_invitation_pending=None)
+        user_disabled_active2 = self.factory.create_user(disabled_at=now, is_invitation_pending=False)
+        user_disabled_pending = self.factory.create_user(disabled_at=now, is_invitation_pending=True)
+
+        # get all enabled
```

Comment:
We should strive to have one logical assertion per test. Could you separate all these to different tests?

Reply:
I decided to put them together because they use the same set of users and are related to the same functionality. Wasn't sure how to do it right :man_shrugging:

Reply:
`make_request_and_check_users` is a step in the right direction. Next I'd move all those `create_user` calls to the fixture level and split each call to `make_request_and_check_users` to its own test, indicating what it is testing ("test_gets_all_disabled_and_pending" etc)

Reply:
Another thing that bothers me - I'd prefer to see a more idiomatic test which actually asserts in the test body. Consider transforming `make_request_and_check_users` so it makes the request and returns the users, then use a set to make sure that the return value equals your expected users. Something like:

```python
expected = Set([ ua1, ua2 ])
actual = Set(make_request("pending=false"))
self.assertEquals(expected, actual)
```

Also, this way you don't have to assert about what _doesn't_ come back.

Reply:
https://github.com/getredash/redash/pull/3400#discussion_r254285338